### PR TITLE
[Generate] Add save mode logits processor to remove nans and infs if necessary

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

It can happen that the output logits of models contain `inf` or even `nan` values. Those values will necessarily lead to errors when using the `sample(...)` or `beam_sample(...)` method. 

This PR adds an optional `InfNanRemoveLogitsProcessor` that - enabled - should remove those values. It should help to fix flaky ci failures like this one: https://app.circleci.com/pipelines/github/huggingface/transformers/21081/workflows/36711d05-4282-4167-88df-59fbda03fe33/jobs/181274

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -789,6 +794,9 @@ def generate(
                 needs to be the target language token.
             forced_eos_token_id (:obj:`int`, `optional`):
                 The id of the token to force as the last generated token when :obj:`max_length` is reached.
+            save_mode (:obj:`bool`, `optional`):
+                Whether to remove possible `nan` and `inf` outputs of the model to prevent the generation method to
+                crash. Note that using ``save_mode`` can slow down generation.
```
Comment (sgugger) [inline src/transformers/generation_utils.py:800]: Why `save_mode`? I don't see how it's linked to `nan` and `inf`.
Also does it default to `None` because it has different defaults depending on something else or because there are three different modes (None, True, False)? Could you elaborate on the doc what happens if it's unset/set to True, set to False?

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -639,6 +641,8 @@ def _get_logits_processor(
             processors.append(ForcedBOSTokenLogitsProcessor(forced_bos_token_id))
         if forced_eos_token_id is not None:
             processors.append(ForcedEOSTokenLogitsProcessor(max_length, forced_eos_token_id))
+        if save_mode is True:
```
Comment (sgugger) [inline src/transformers/generation_utils.py:644]:
```suggestion
        if save_mode:
```

**Review by sgugger — COMMENTED** [review]
I'm a bit perplexed by `save_mode`. Why this name? Also why does it default to None when from the code I only see it as a True/False flag (and it should therefore default to False and not be Optional[bool])?

The rest is very clean, thanks!

**Review by LysandreJik — APPROVED** [review]
LGTM, great tests and clean implementation!

I agree with Sylvain on the `save_mode`, maybe something like `remove_invalid_values` or `remove_overflowing_values` would be better?

File: src/transformers/generation_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py]: Changed it so that it defaults back to the parameter as defined in the config. Also agree that `save_mode` is not a good name -> changed it to `remove_invalid_values` as suggested by @LysandreJik
